mathics/core/symbols.py

```diff
@@ -288,6 +288,15 @@ def get_sort_key(self, pattern_sort=False):
         else:
             raise NotImplementedError
 
+    def has_changed(self, definitions) -> bool:
+        """
+        Used in Expression.evaluate() to determine if we need to reevaluation
+        an expression.
+        """
+        # It would be nice to make the default "False" here and have Symbols
```

Comment:
I think it should be false.

Reply:
In one situation, testing if an expression or atom has_changed` is used in rebuilding a cache.  I addressed that particular place, but I think there are other places that need to be fixed before we can make this False.

I am kind of disgusted at how the caching which should have been limited to perhaps compound expressions has invaded everything so everything has to coddle to the cache. (And if the cache were _that_ good we wouldn't be having the performance problems we have). 

It should not be mandatory for an Atom to provide caching when it is nonsensical. Rather it should be up to the evaluator to see whether there is a cache. I have only looked at and changed one place. 

As too often is the case there is too much mess and I have too little time for me to try to address this in one go.

This is basically what I have for this week. And I _still_ haven't gotten to adding a property on the elements to indicate that all of them never need reevaluation.

Reply:
@TiagoCavalcante @mmatera looking over the single use of `has_changed()` it looks like a more appropriate name would be `needs_evaluation` (in order to set some sort of "value" field?).

For certain numbers like Integers and Rational this could be done initially. I think that's why we have `needs_evaluation` set False for a number of things like String. 

If this is the case, I'd like (next week) to rename this method. 

Comments?